include/expression.hpp

```diff
@@ -0,0 +1,203 @@
+#ifndef ADJACENT_EXPRESSION_HPP
+#define ADJACENT_EXPRESSION_HPP
+
+#include <memory>
+#include <string>
+
+class Expr;
+
+template <class T>
+class Param
+{
+private:
+	T m_value;
+
+public:
+	std::string m_name;
+	bool m_reduceable = true;
+	bool m_changed = false;
+	std::shared_ptr<Expr> m_expr;
+
+	Param() = default;
+	Param(const std::string& name, bool reduceable = true);
+	Param(const std::string& name, double value);
+
+	void set_value(const T& other);
+	T value() const;
+
+	std::shared_ptr<Expr> expr();
+
+	bool operator==(const Param& other) const;
+};
+
+template <class T>
```

Comment:
I prefer inline declaration of members for templates since it will be inside header anyway